telemetry(amazonq): attach stack trace in onCodeGeneration failures

siakmun-aws opened the pull request:
## Problem
To further find issues in onCodeGeneration, need to attach stack trace when sending metrics

## Solution
Attach stack trace when sending metrics

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be squash-merged at release time.
- Your code changes must meet the guidelines in [CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

neilk-aws commented on packages/core/src/test/amazonqFeatureDev/controllers/chat/controller.test.ts:
```diff
@@ -518,7 +522,13 @@ describe('Controller', () => {
+                        metricResult,
+
+                        // Stack trace should include the name of the test framework
+                        sinon.match((str) => typeof str === 'string' && str.includes('Mocha'))
```
Can we assert the useful part of the log message is actually present? This does not guarantee that the stack trace is actually empty.

neilk-aws commented on packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts:
```diff
@@ -1009,4 +1019,55 @@ export class FeatureDevController {
+        const seenExceptions = new Set<Error>()
+        const lines: string[] = []
+
+        function printExceptionDetails(err: Error, prefix: string = '') {
```
Can we add a recursion limit to this function to avoid any large or blocking exceptions that lack a good base case?

neilk-aws commented on packages/core/src/test/amazonqFeatureDev/controllers/chat/controller.test.ts:
```diff
@@ -519,7 +523,11 @@ describe('Controller', () => {
+                    sendMetricDataTelemetrySpy.calledWith(
+                        MetricDataOperationName.EndCodeGeneration,
+                        metricResult,
+                        sinon.match((str) => typeof str === 'string' && str.includes('stack trace: '))
```
Can we assert more of the contents of the stack trace? Otherwise, there is nothing else really that is testing the method that you added

justinmk3 reviewed (changes requested):
This is lacking an acceptable explanation of what the purpose is, and how it works. Did you test this with a production webpack'd build? Nodejs stacktraces in the production build are usually obfuscated.

The docstring of the function does not give useful insights. 

The function is non-trivial, and has no test coverage. 

Assuming the function works, it belongs in a core util module, not randomly in the middle of a feature module.

siakmun-aws commented:
> This is lacking an acceptable explanation of what the purpose is, and how it works. Did you test this with a production webpack'd build? Nodejs stacktraces in the production build are usually obfuscated.
> 
> The docstring of the function does not give useful insights.
> 
> The function is non-trivial, and has no test coverage.
> 
> Assuming the function works, it belongs in a core util module, not randomly in the middle of a feature module.

Here is some background on the purpose of this PR. The previous (main branch) client has been able to send telemetry in the event of code generation failure. However, any uncaught errors, either toolkit or runtime error etc., will only result in sending a "Fault" telemetry. We want to be able to diagnose such errors, and therefore the stack trace is needed. TS obfuscation is expected since non-toolkit errors such as runtime errors are not obfuscated. Uncaught toolkit error names are obfuscated, but through the function names (not obfuscated) in the stack trace combined with the code line number and toolkit version, the cause can still be identified.

Docstring: modified since last commit to give more details
Test: added
File place: changed to shared/error.ts

justinmk3 commented on packages/core/src/shared/errors.ts:
```diff
@@ -368,6 +368,57 @@ export function getTelemetryResult(error: unknown | undefined): Result {
 
+export class SafeMessageToolkitError extends ToolkitError {
+    public constructor(message: string, info: ErrorInformation = {}) {
+        super(message, info)
```
what is going on here? why was this new base class introduced? what is the intended meaning of "safe message" ?

siakmun-aws replied:
Safe message means the error message is safe to send to telemetry (no critical information). This was introduced to allow error classes which extends it can be sent to telemetry with its error message.
But this PR is not ready yet, so I will draft it for now.

justinmk3 replied:
That doesn't sound like the right direction. We already have `scrubNames` https://github.com/aws/aws-toolkit-vscode/blob/7d5a339c4334d6ac4bb3123b81b0073286b82dfe/packages/core/src/shared/errors.ts#L378

which removes PII from strings. And that is used by `getTelemetryReasonDesc`, which already is used automatically.

Adding new, special classes is not the right way to approach this.
